cmd/branchprotector/protect.go

```diff
@@ -133,6 +135,14 @@ func main() {
 		enabled:                o.githubEnablement.EnablementChecker(),
 	}
 
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+	go func() {
+		s := <-interrupt
+		logrus.Errorf("Received an interrupt: %s, cancelling...", s)
+		os.Exit(1)
```

Comment:
I don't think that it's safe to just exit here. Instead, we need to gracefully stop all the other goroutines. IIRC the interrupts library is already handling this case, perhaps you can take a look.

Reply:
Thanks for the comment @droslean. I'm not a Go expert, but branchprotector is a one-shot command and not a server/service. All of the prow components that use interrupts are running an actual server, and they have context etc. Using Interrupts doesn't seem to fit here, and since branchprotector isn't a server, I'm not seeing the harm of Exiting right away? I don't see any other prow one-shot command use Interrupts either.

Can you point to relevant example to help me out?

Reply:
[Here](https://github.com/kubernetes-sigs/prow/pull/144/files#diff-58095a12a640f817a91a9590a6b391dcff66b3e2560ce85713272c932e687a8dR146) we are spawning a go routine to configure the branches, and in this [line](https://github.com/kubernetes-sigs/prow/pull/144/files#diff-58095a12a640f817a91a9590a6b391dcff66b3e2560ce85713272c932e687a8dR147) another operation starts which both share multiple channels to communicate. Instead of force exiting, we should ensure that the channels are closed and all go routines stopped, we should gather the errors and then exit.

Reply:
TBH @droslean I tend to agree that there's little to gain for a one-shot command implementing graceful shutdown...

Reply:
@petr-muller I agree that force-exiting can result in not updating some repositories of a specific org, which is not a big deal, but eating an error in the `configureBranches()` go routine, shouldn't be wise. Ref https://github.com/kubernetes-sigs/prow/blob/main/cmd/branchprotector/protect.go#L185 that is being consumed in https://github.com/kubernetes-sigs/prow/blob/main/cmd/branchprotector/protect.go#L139. In addition, implementing a graceful exiting should be trivial if the context can be utilized appropriately.